### ISSUE TITLE
Roslyn fails to verify [NotNullWhen] contracts in methods returning Nullable<bool>

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#57649]
**Version Used**: 
```
Branch main (5 Nov 2021)
Latest commit 44f0576 by Joey Robichaud:
Revert to VS2019 for official build
```

**Steps to Reproduce**:

Compile and run the following code
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

class C
{
  public static void Main()
  {
    if (M1(out var x) ?? false)
    {
      x.ToString();
    }
  }

  static bool? M1([NotNullWhen(true)] out string? x)
  {
    x = null;
    return true;
  }
}
```

**Expected Behavior**:
`CS8762: Parameter 'x' must have a non-null value when exiting with 'true'.` reported for `return true;` in `M1`

**Actual Behavior**:
No warnings at all in the program above. It crashes at runtime with a `NullReferenceException`

**Comment by RikkiGibson** [comment]
@jcouv did we inadvertently do the "call-site" side of this without doing the "declaration" side of it?

**Comment by jcouv** [comment]
Conditional nullable attributes are only expected to work on `bool`, not `Nullable<bool>`.
https://github.com/dotnet/csharplang/blob/main/meetings/2019/LDM-2019-05-15.md#conditional-postconditions

The bug here is actually on the call-site. We are not able to analyze the call-site because we only track one or two states (unconditional or conditional respectively), but `Nullable<bool>` involves more than two states...

**Comment by TessenR** [comment]
@jcouv Does it mean that the following code should have a warning because the expression `dictionary?.TryGetValue("", out var x)` is of type `Nullable<bool>`?

```cs
#nullable enable
using System.Collections.Generic;

class C
{
  public void M(Dictionary<string, string>? dictionary)
  {
    if (dictionary?.TryGetValue("", out var x) ?? false)
    {
      x.ToString(); // currently no warning
    }
  }
}
```

**Comment by jcouv** [comment]
@TessenR No, that scenario will continue to work. `TryGetValue` returns a bool. We have a special treatment for this whole expression pattern (ie. `x?.y ?? false`) as part of [improved definite assignment](https://github.com/dotnet/csharplang/issues/4465).

**Comment by sharwell** [comment]
It seems like one solution here is to report a diagnostic for use of `NotNullWhen`/`MaybeNullWhen` on non-`bool`-returning methods.

**Comment by RikkiGibson** [comment]
#36073
